File: EEG101/android/app/src/main/java/com/eeg_project/components/classifier/EpochBuffer.java
package com.eeg_project.components.classifier;

import android.util.Log;

import com.eeg_project.components.signal.CircularBuffer;

import java.util.ArrayList;

/**
 * A circular buffer that can register listeners to be sent epochs of data at a given interval
 */

public class EpochBuffer extends CircularBuffer {

    private int epochInterval;
    private ArrayList<BufferListener> listeners = new ArrayList<BufferListener>();

    // ------------------------------------------------------------------------
    // Constructor

    public EpochBuffer(int bufferLength, int nChannels, int interval) {
        super(bufferLength, nChannels);
        epochInterval = interval;
    }

    public void addListener(BufferListener listener) {
        listeners.add(listener);
    }

    @Override
    public void update(double[] newData) {

        for(int i = 0; i < nbCh; i++) {
            buffer[index][i] = newData[i];
        }
        index = (index + 1) % bufferLength;
        if (index % epochInterval == 0) {
            try {
                for (BufferListener listener : listeners) {
                    listener.getEpoch(extractTransposed(bufferLength));
                }
            } catch (Exception e) {
                Log.w("Buffer", e);
            }
        }
        pts++;
    }

}

File: EEG101/ios/Source/Classifier Module/EpochBuffer.swift
import Foundation

protocol BufferListener {
    func getEpoch(buffer: [[Double]])
}

class EpochBuffer: CircularBuffer {
    private let epochInterval: Int
    private var listeners: [BufferListener]
    
    init(bufferLength: Int, nChannels: Int, epochInterval: Int) {
        self.epochInterval = epochInterval
        self.listeners = [BufferListener]()
        
        super.init(bufferLength: bufferLength, nChannels: nChannels)
    }
    
    func addListener(_ listener: BufferListener) {
        listeners.append(listener)
    }
    
    override func update(with newData: [Double]) {
        super.update(with: newData)
        
        let extraction = extractTransposed(sampleCount: bufferLength)
        listeners.forEach {
            $0.getEpoch(buffer: extraction)
        }
    }
}
